File: src/main/java/pl/qprogramming/devinbox/shared/utils/EmailUtils.java
package pl.qprogramming.devinbox.shared.utils;

import lombok.experimental.UtilityClass;

@UtilityClass
public class EmailUtils {

    /**
     * Masks the given email address by showing only the first character of the local part
     * and the first character of the domain part, replacing the rest with asterisks
     * @param email the email address to be masked
     * @return the masked email address
     */
    public static String maskEmail(String email) {
        if (email == null || email.isEmpty()) {
            return "***";
        }

        int atIndex = email.indexOf('@');
        if (atIndex <= 0) {
            return "***";
        }

        String localPart = email.substring(0, atIndex);
        String domain = email.substring(atIndex + 1);

        String maskedLocal = localPart.charAt(0) + "***";

        int dotIndex = domain.indexOf('.');
        String maskedDomain = dotIndex > 0
            ? domain.charAt(0) + "***" + domain.substring(dotIndex)
            : domain.charAt(0) + "***";

        return maskedLocal + "@" + maskedDomain;
    }
}

File: DevInbox.Web/Common/Utils/EmailUtils.cs
namespace DevInbox.Web.Common.Utils;

public static class EmailUtils
{
    /// <summary>
    /// Masks the given email address by showing only the first character of the local part
    /// and the first character of the domain part, replacing the rest with asterisks.
    /// </summary>
    /// <param name="email">The email address to be masked.</param>
    /// <returns>The masked email address.</returns>
    public static string MaskEmail(string? email)
    {
        if (string.IsNullOrEmpty(email))
        {
            return "***";
        }

        var atIndex = email.IndexOf('@');
        if (atIndex <= 0)
        {
            return "***";
        }

        var localPart = email[..atIndex];
        var domain = email[(atIndex + 1)..];

        var maskedLocal = $"{localPart[0]}***";

        var dotIndex = domain.IndexOf('.');
        var maskedDomain = dotIndex > 0
            ? $"{domain[0]}***{domain[dotIndex..]}"
            : $"{domain[0]}***";

        return $"{maskedLocal}@{maskedDomain}";
    }

    /// <summary>
    /// Normalizes an email address by trimming whitespace and converting to lowercase.
    /// Returns null if the input email is null.
    /// </summary>
    public static string? NormalizeEmail(string email)
    {
        return email?.Trim().ToLower();
    }
}
